From 39621631bb5e0f550078d24cbd148076e4279342 Mon Sep 17 00:00:00 2001
From: Stefan Lange <s.lange@gateware.de>
Date: Wed, 12 Oct 2016 13:35:41 +0200
Subject: [PATCH] TQMT1040: automatic reprogramming of ethernet phy if SerDes
 configured for QSGMII

The 88E1340 ethernet phy on the STKTxxxx starterkit can operate in SGMII
or QSGMII mode. As default out-of-reset value, SGMII is set in the phys
registers.
If the TQMT1040 SerDes configuration is configured for QSGMII, this
patch automatically reprograms the ethernet phy for QSGMII operation.
---
 board/tqc/tqmt1042/eth.c   | 20 ++++++++++++++++++--
 include/configs/TQMT1042.h |  1 +
 2 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/board/tqc/tqmt1042/eth.c b/board/tqc/tqmt1042/eth.c
index 4efa001..e17a4cd 100644
--- a/board/tqc/tqmt1042/eth.c
+++ b/board/tqc/tqmt1042/eth.c
@@ -14,6 +14,8 @@
 #include <asm/fsl_dtsec.h>
 #include <phy.h>
 
+static int eth_phy_qsgmii_reprogram = 0;
+
 /*
  * Set RGMII delay values in ethernet phys
  * located on the STKT104x starterkit
@@ -93,6 +95,13 @@ int board_phy_config(struct phy_device *phydev)
 		phy_write(phydev,MDIO_DEVAD_NONE,0x1d,0x0000);
 	}
 
+	if ((phydev->addr == CONFIG_SYS_QSGMII1_PHY_ADDR) && (eth_phy_qsgmii_reprogram == 1)) {
+		printf("Eth:   reprogramming 88E1340 ethernet phy to QSGMII\n");
+                phy_write(phydev,MDIO_DEVAD_NONE,MII_88E1340_PAGEAD,0x06);
+                phy_write(phydev,MDIO_DEVAD_NONE,0x14,0x8200);
+                phy_write(phydev,MDIO_DEVAD_NONE,MII_88E1340_PAGEAD,0x00);
+	}
+
 	return 0;
 }
 
@@ -140,8 +149,15 @@ int board_eth_init(bd_t *bis)
 			fm_info_set_phy_address(i, phy_addr);
 			break;
 		case PHY_INTERFACE_MODE_QSGMII:
-			/* QSGMII unsupported on T1042 */
-			fm_info_set_phy_address(i, 0);
+			/* QSGMII unsupported on T1042, supported on T1040
+			 * TODO: QSGMII currently errorous on hardware
+			 * implement after hardware has been fixed
+			 * only automatic reprogramming of the phy
+			 * from SGMII to QSGII
+			 * has been implemented so far */
+			printf("Eth:   configuring FM1_DTSEC%i as QSGMII\n", i+1);
+			eth_phy_qsgmii_reprogram = 1;
+			fm_info_set_phy_address(i, phy_addr);
 			break;
 		case PHY_INTERFACE_MODE_NONE:
 			fm_info_set_phy_address(i, 0);
diff --git a/include/configs/TQMT1042.h b/include/configs/TQMT1042.h
index 2d648e3..cb9462b 100644
--- a/include/configs/TQMT1042.h
+++ b/include/configs/TQMT1042.h
@@ -535,6 +535,7 @@
 #define CONFIG_SYS_SGMII4_PHY_ADDR		0x01
 #define CONFIG_SYS_RGMII1_PHY_ADDR		0x0E
 #define CONFIG_SYS_RGMII2_PHY_ADDR		0x05
+#define CONFIG_SYS_QSGMII1_PHY_ADDR             0x1C
 
 #define CONFIG_MII		/* MII PHY management */
 #define CONFIG_ETHPRIME		"FM1@DTSEC4"
-- 
1.9.1

